web/packages/teleterm/src/mainProcess/clusterStore/clusterStore.ts

```diff
@@ -0,0 +1,180 @@
+      for (let d of draft.values()) {
+        if (routing.belongsToProfile(uri, d.uri)) {
+          draft.delete(d.uri);
+        }
+      }
+    });
+  }
+
+  /**
+   * Synchronizes the root clusters.
+   * Does not make a network call, only reads profiles from the disk.
+   */
+  async syncRootClusters(): Promise<void> {
+    const { response } = await this.tshdClient.listRootClusters({});
+    await this.update(draft => {
+      draft.clear();
+      response.clusters.forEach(cluster => {
+        draft.set(cluster.uri, cluster);
+      });
+    });
+  }
+
+  /**
+   * Synchronizes the root cluster.
+   * Makes a network call to get cluster details and its leaf clusters.
+   */
+  async sync(uri: RootClusterUri): Promise<void> {
+    try {
+      const { cluster, leafs } = await getClusterAndLeafs(this.tshdClient, uri);
+
+      await this.update(draft => {
+        draft.set(cluster.uri, cluster);
+        leafs.forEach(leaf => {
+          draft.set(leaf.uri, leaf);
+        });
+      });
+    } catch (error) {
+      await this.update(draft => {
+        const cluster = draft.get(uri);
+        if (cluster) {
+          // TODO(gzdunek): We should rather store the cluster synchronization status,
+          // so the callsites could check it before reading the field.
+          // The workaround is to update the field in case of a failure,
+          // so the places that wait for showResources !== UNSPECIFIED don't get stuck indefinitely.
+          cluster.showResources = ShowResources.ACCESSIBLE_ONLY;
+        }
+      });
+      throw error;
+    }
+  }
+
+  getRootClusters(): Cluster[] {
+    return this.state
+      .values()
+      .toArray()
+      .filter(c => !c.leaf);
+  }
+
+  getState(): State {
+    return this.state;
+  }
+
+  /**
+   * Registers a sender.
+   * Sends the full current state as the initial message, and then patches.
+   * This method blocks until the sender is disposed.
+   */
+  async useSender(sender: AwaitableSender<ClusterStoreUpdate>): Promise<void> {
+    this.logger.info('Sender added');
+    this.senders.add(sender);
+    await sender.send({ kind: 'state', value: this.state });
+    await sender.whenDisposed();
+    this.senders.delete(sender);
+    this.logger.info('Sender removed');
+  }
+
+  private async update(producer: Producer<State>): Promise<void> {
+    const [state, patches] = produceWithPatches(this.state, producer);
+    this.state = state;
+    await Promise.all(
+      this.senders
+        .values()
+        .map(sender => sender.send({ kind: 'patches', value: patches }))
+    );
+  }
+}
+
+async function getClusterAndLeafs(tshdClient: TshdClient, uri: RootClusterUri) {
+  const resolved = await Promise.all([
+    tshdClient.getCluster({
+      clusterUri: uri,
+    }),
+    tshdClient.listLeafClusters({
+      clusterUri: uri,
+    }),
+  ]);
+
+  return {
+    cluster: resolved[0].response,
+    leafs: new Map(resolved[1].response.clusters.map(c => [c.uri, c])),
```

Comment:
Do we need to put them in a map if we don't end up using any of the features of the map?